### DISCUSSION TITLE
Next 14: Specified module format (CommonJs) is not matching the module format of the source code (EcmaScript Modules)

### DISCUSSION THREAD

**Discussion by dominik-sfl** [pmndrs/jotai#2579] [Bug report]
In my newly installed next.js app (14.2.3), using `create-t3-app`, I get the following error when I import jotai's Provider & wrap it around its children:

> ⨯ ModuleBuildError: ./node_modules/jotai/esm/react.mjs/proxy.js
> Specified module format (CommonJs) is not matching the module format of the source code (EcmaScript Modules)
> The CommonJs module format was specified in the package.json that is affecting this source file or by using an special extension, but Ecmascript import/export syntax is used in the source code.
> The module was automatically converted to an EcmaScript module, but that is in conflict with the specified module format. Either change the "type" field in the package.json or replace EcmaScript import/export syntax with CommonJs syntas in the source file.
> In some cases EcmaScript import/export syntax is added by an transform and isn't actually part of the source code. In these cases revisit transformation options to inject the correct syntax.

Simply by doing this:

```
<html
      lang="en"
      className={`font-sans ${GeistMono.variable} ${noto_sans.variable}`}
    >
        <body>
          <TRPCReactProvider>
            <Provider>
              {children}
            </Provider>
          </TRPCReactProvider>
        </body>
    </html>
```

TS Config (default t3 tsconfig)

```
{
  "compilerOptions": {
    /* Base Options: */
    "esModuleInterop": true,
    "skipLibCheck": true,
    "target": "es2022",
    "allowJs": true,
    "resolveJsonModule": true,
    "moduleDetection": "force",
    "isolatedModules": true,

    /* Strictness */
    "strict": true,
    "noUncheckedIndexedAccess": true,
    "checkJs": true,

    /* Bundled projects */
    "lib": ["dom", "dom.iterable", "ES2022"],
    "noEmit": true,
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "jsx": "preserve",
    "plugins": [{ "name": "next" }],
    "incremental": true,

    /* Path Aliases */
    "baseUrl": ".",
    "paths": {
      "~/*": ["./src/*"]
    }
  },
  "include": [
    ".eslintrc.cjs",
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx",
    "**/*.cjs",
    "**/*.js",
    ".next/types/**/*.ts"
  ],
  "exclude": ["node_modules"]
}
```

Anyone know what's going on here? :/

**Comment by dai-shi** [comment]
Hm, I've never seen such an error message. Would you be able to create a reproduction with StackBlitz?

**Reply by cjimmy** [reply]
Here's a minimal repro with the Next.js starter (App Router). I think the issue has something to do with turbo pack. No issue on jotai 2.8.3 when running vanilla `next dev`. Issue appears when running `next dev --turbo`
https://codesandbox.io/p/devbox/funny-sun-zpj5hv?file=%2Fapp%2Fpage.tsx%3A4%2C28

(sorry it's not StackBlitz. I'm used to codesandbox and got stuck with StackBlitz!)

**Reply by dai-shi** [reply]
sorry, i'm not very familiar with turbo. 

See also: https://github.com/pmndrs/jotai/discussions/2579#discussioncomment-9718694

Can you check if this works for you?
https://ci.codesandbox.io/status/pmndrs/jotai/pr/2555/builds/506960
See "Local Install Instructions" ☝️

**Reply by matzkoh** [reply]
I used pkg.csb.dev to identify which [commit](https://github.com/pmndrs/jotai/commits/main) introduced this error. Everything worked fine up until before #2548.

**Reply by cjimmy** [reply]
> sorry, i'm not very familiar with turbo.

[Turbopack](https://nextjs.org/docs/architecture/turbopack) is a webpack replacement, used by Nextjs, but not yet generally available. However, it appears that it will be the default bundler in the near future for Nextjs. So, this particular issue may not be worth addressing now, but it will be a larger issue within the next year (my guess).

> See also: https://github.com/pmndrs/jotai/discussions/2579#discussioncomment-9718694

I can confirm that downgrading to jotai 2.8.0 works with turbopack

> Can you check if this works for you?

I still get the same issue on the minimal repro when I install #2555 and run `pnpm dev --turbo`

<img width="651" alt="Screenshot 2024-06-17 at 12 34 18 PM" src="https://github.com/pmndrs/jotai/assets/700885/cffc59d5-b3c2-404b-8afb-b46fe6e63203">

**Reply by dai-shi** [reply]
Thanks a lot, guys. I didn't expect #2548 changed something significant.
Can you dig it a deeper and find what exactly causes the problem?

Perhaps adding "commonjs"? https://github.com/pmndrs/jotai/pull/2548/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R5

**Reply by matzkoh** [reply]
@dai-shi

I installed version 2.8.3 and tested it multiple times by adding and removing `"type": "commonjs",` from `node_modules/jotai/package.json`. It worked when I removed it!

**Reply by dai-shi** [reply]
Oh, that's very very great to know!

Now I wonder if it's turbo's issue. Would anyone like to report it to them?

The rationale is publint warns it: https://publint.dev/rules#use_type

**Reply by saoudi-h** [reply]
> Oh, that's very very great to know!
> 
> Now I wonder if it's turbo's issue. Would anyone like to report it to them?
> 
> The rationale is publint warns it: https://publint.dev/rules#use_type

Hi,

I opened an issue on the Next.js repo [issue:74062](https://github.com/vercel/next.js/issues/74062)

The Next.js team has already proposed a fix, and it might resolve this incompatibility

**Reply by dai-shi** [reply]
Thanks!

**Reply by saoudi-h** [reply]
No problem!

**Comment by danimrangelb** [comment]
Same here! @dominik-sfl are you using turbo?

**Comment by sansan-sakura** [comment]
I got the same error using turbo.  @dominik-sfl @danimrangelb Has anyone solved this issue?

**Comment by mikko-kohtala** [comment]
We got the same error when upgrading from `2.8.0` to `2.8.1`. Happens also with the latest `2.8.3`. We are using turbo.

Kind of big project, but most of the dependencies are at the latest. Some notable deps:
* "next": "14.2.3",
* "next-auth": "5.0.0-beta.19",
* trpc @ `11.0.0-rc.390`

Happy to answer any questions.

**Reply by dai-shi** [reply]
> when upgrading from `2.8.0` to `2.8.1`

That's great info.
v2.8.0 works, and v2.8.1 breaks it, right?

https://github.com/pmndrs/jotai/releases/tag/v2.8.1
Unfortunately, I'm not sure which change can be related.

https://github.com/pmndrs/jotai/pull/2563/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R147
Maybe, the update of esbuild can be something?

**Comment by dai-shi** [comment]
I made a draft PR to remove the type field in package.json, but we don't plan to merge it as of now: https://github.com/pmndrs/jotai/pull/2626

**Reply by malixswoop** [reply]
Thanks, downgrading to 2.8.0

**Reply by theogravity** [reply]
Is there a reason to have it there? Trying 2.9.3 and have the same issue (using in a pnpm monorepo with turbo + next.js using --turbo flag).

Edit: I see. I would personally favor compatibility over startup performance.

**Reply by dai-shi** [reply]
https://publint.dev/rules#use_type

> [Node.js v21.1.0](https://nodejs.org/en/blog/release/v21.1.0) adds a new --experimental-detect-module, which can be used to automatically run ES modules when ESM syntax can be detected. Node.js hopes to make detection enabled by default in the future. Detection increases startup time, so Node is encouraging everyone — especially package authors — to add a type field to package.json, even for the default "type": "commonjs".

**Reply by gpxl-dev** [reply]
I also hit this and would prefer not to have to downgrade.

**Reply by dai-shi** [reply]
For now, I keep updating #2626 if that helps.

Does anybody know if Next 15 rc does something different?

**Reply by theogravity** [reply]
It does not do anything differently - we're on next 15 RC

**Reply by dai-shi** [reply]
That's unfortunate. Did anyone file an issue in the next repo?

**Reply by lamualfa** [reply]
> For now, I keep updating #2626 if that helps.
> 
> Does anybody know if Next 15 rc does something different?



@dai-shi Error when using jotai in Next.js `15.0.4-canary.52` & `15.1.1-canary.2` _(I've tried 2 versions):_

```bash
⨯ ./node_modules/.pnpm/jotai@2.10.3_@types+react@19.0.1_react@19.0.0/node_modules/jotai/esm/react.mjs/proxy.js
Specified module format (CommonJs) is not matching the module format of the source code (EcmaScript Modules)
The CommonJs module format was specified in the package.json that is affecting this source file or by using an special extension, but Ecmascript import/export syntax is used in the source code.
The module was automatically converted to an EcmaScript module, but that is in conflict with the specified module format. Either change the "type" field in the package.json or replace EcmaScript import/export syntax with CommonJs syntas in the source file.
In some cases EcmaScript import/export syntax is added by an transform and isn't actually part of the source code. In these cases revisit transformation options to inject the correct syntax.



./node_modules/.pnpm/jotai@2.10.3_@types+react@19.0.1_react@19.0.0/node_modules/jotai/esm/react/utils.mjs/proxy.js
Specified module format (CommonJs) is not matching the module format of the source code (EcmaScript Modules)
The CommonJs module format was specified in the package.json that is affecting this source file or by using an special extension, but Ecmascript import/export syntax is used in the source code.
The module was automatically converted to an EcmaScript module, but that is in conflict with the specified module format. Either change the "type" field in the package.json or replace EcmaScript import/export syntax with CommonJs syntas in the source file.
In some cases EcmaScript import/export syntax is added by an transform and isn't actually part of the source code. In these cases revisit transformation options to inject the correct syntax.
```

**Reply by dai-shi** [reply]
It looks like a different error message we've been discussing.
I'm not sure but `/proxy.js` has `.js` extension, which might be a problem. If it were `.mjs`, the "type" field doesn't matter.

**Comment by theogravity** [comment]
I was able to edit out `"type": "commonjs"` using the pnpm package patcher:

https://pnpm.io/cli/patch

Doing this fixed it.

**Reply by mycolaos** [reply]
You edited jotai package.json?

**Reply by guseggert** [reply]
Similarly for NPM I used [patch-package](https://www.npmjs.com/package/patch-package).  

* Go into `node_modules/jotai/package.json`, remove the `"type": "commonjs"` line
* Run `npx patch-package jotai --exclude nothing` to create the patch file
* Add `"postinstall": "patch-package"` to `package.json`

**Comment by cchin25** [comment]
Just for anyone googling, just ran into this with next `14.2.4` and jotai `2.10.1` when using the `--turbo` flag when running dev.

**Reply by jonathanwilke** [reply]
Yeah same for me with the latest version of Next.js (15.0.0). Were you able to resolve this with the --turbo flag?

**Reply by sleeporsmile** [reply]
> Yeah same for me with the latest version of Next.js (15.0.0). Were you able to resolve this with the --turbo flag?

same issue

**Reply by cchin25** [reply]
> Yeah same for me with the latest version of Next.js (15.0.0). Were you able to resolve this with the --turbo flag?

I was unable to, but tbh I didn't put much effort into it. At the time I just ended up removing the --turbo flag. Ultimately for unrelated reasons I switched to Zustand.

**Reply by songkeys** [reply]
I removed the --turbo flag and it's still the same.

**Comment by SebasScript** [comment]
Also ran into same issue spinning up a new Next.js 15.0.2 using Turbo in dev mode. 
Downgrading Jotai to  2.8.0 as mentioned earlier in the thread gets around the type error as workaround

**Comment by songkeys** [comment]
To fix this using pnpm:

Prepare for patching:

```bash
pnpm patch jotai
```

Open package.json:

```bash
open node_modules/.pnpm_patches/jotai@2.10.1/package.json
```

Remove this line (L5):

```diff
{
  "name": "jotai",
  "description": "👻 Primitive and flexible state management for React",
  "private": false,
-  "type": "commonjs",
  "version": "2.10.1",
  "main": "./index.js",
```

Path this commit:

```bash
pnpm patch-commit 'node_modules/.pnpm_patches/jotai@2.10.1'
```

Now it should work though I don't know why.

See: https://pnpm.io/cli/patch

**Comment by 8spw** [comment]
That happens when you are using jotai in SSR, make sure put `"use client"` in top of component

**Reply by lubieowoce** [reply]
note that this is not required, `jotai` already includes a "use client" in their code -- this is just a workaround that happens to sidestep the next.js bug (specifically, the bug shouldn't be triggered if `jotai` is only ever imported from client modules). i'm hoping we'll have a fix merged/released soon, and then your `"use client"` may not be necessary anymore.

**Comment by miruji** [comment]
The only thing that helped me was using dynamic loading without ssr, which guarantees full operation on the client

**Comment by saoudi-h** [comment]
Hi everyone,

Just a quick update: this issue has been fixed in Next.js 15.1.4. No more workarounds or patches are needed.

Details here: [Next.js PR #74608](https://github.com/vercel/next.js/pull/74608).

**Reply by lubieowoce** [reply]
I was going to post it here too, thanks :)
For the folks who are still on v14, we've also backported this fix (and others) in [14.2.23](https://github.com/vercel/next.js/releases/tag/v14.2.23).
